runtime/collect/collect.cpp

```diff
@@ -380,14 +374,13 @@ bool store_map_for_gc(void **roots, map *ptr) {
   *offset = 16;
   auto *child = (map *)(hdr + 2);
   *child = std::move(*ptr);
-  *roots = mem;
-  return true;
+  *roots = child;
```

Comment:
Could you explain the reason we store the `child` pointer here instead of the `mem` pointer as it was before? There is a similar change in the corresponding functions for sets, lists, and rangemaps.

Reply:
okay, yes, I can explain. So, the garbage collection roots in this case are collections that were stored on the alwaysgcspace. In order to allow them to survive garbage collection, which alwaysgcspace objects cannot normally do, we are promoting them in this function to objects on the youngspace. This requires putting them as the child of a regular symbol (which is in this case a special symbol created for this specific purpose), and giving that regular symbol a block header. The `mem` pointer points to that block header. Previously, `is_block` was set to true in this case, which caused the garbage collector to know that the pointer pointed to a symbol and not a collection, and caused the `load_for_gc` function to know that it had to offset the pointer to get back to the collection.

However, logic already exists in the garbage collector (due to the most recent set of bug fixes) which allows the garbage collector to handle gc roots of sort map/list/set. It does this by using a special offset word embedded in blocks to be able to find the root of the symbol that the map/list/set is a derived pointer of, and then migrate the entire block to the tospace. This is the logic that is now being used in the case that was previously using `is_block`. In essence, instead of telling the gc directly "no, this is not actually a collection, it's a block containing a collection", we treat the root like any other derived pointer, allowing the collector to use the offset embedded in the block to find the root pointer of the allocation.

However, this means that this function needs to return a derived pointer to the collection, rather than the base pointer of the allocation. That is the change being made in this method; instead of returning the base pointer of the object, we are returning the derived pointer to the child of the object that will then be used by the gc algorithm to reconstruct the base pointer.

Reply:
Makes sense. I am approving since everything looks good.